src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs

```diff
@@ -380,6 +357,53 @@ private static bool SymbolsAreCompatibleCore(ISymbol symbol, ISymbol newSymbol,
             return false;
         }
 
+        private static bool CompareAcrossSemanticModels(ISymbol symbol, ISymbol newSymbol)
+        {
+            // SymbolEquivalenceComparer performs Location equality checks for locals, labels, range-variables and local
+            // functions. As we are comparing symbols from different semantic models, locations will differ. Hence
+            // perform minimal checks for these symbol kinds.
+
+            if (symbol.Kind != newSymbol.Kind)
+                return false;
+
+            if (symbol is ILocalSymbol localSymbol && newSymbol is ILocalSymbol newLocalSymbol)
+            {
+                return newSymbol.IsImplicitlyDeclared == symbol.IsImplicitlyDeclared &&
+                       symbol.Name == newSymbol.Name &&
+                       CompareAcrossSemanticModels(localSymbol.Type, newLocalSymbol.Type);
```

Comment:
❔ Did this change (from `newLocalSymbol.Type.Equals(localSymbol.Type)`) fix a bug? If so, I'm curious which one. Feel free to put the answer in reply to this comment and not in code, since a test will already be covering regressions.

Reply:
the primary issue here is in terms of us calling into this codepath for a speculative semantic model.  In that scenario i wanted to make it so that we never depended on direct-equals of symbols, but instead went through the symbolequivcomparer.